### PR TITLE
Sqrt issue

### LINKED ISSUE #7905: sqrt without std:: in Math/GenVector/PxPyPzE4D.h
Hi,

While trying to compile some code which uses ROOT::Math::PxPyPzMVector, I run into
```
In file included from v6-22-06/install/include/Math/Vector4D.h:12:
v6-22-06/install/include/Math/GenVector/PxPyPzE4D.h:154:17: error: call to function 'sqrt' that is neither visible in the template definition nor found by argument-dependent lookup
         return sqrt(mm);
                ^
v6-22-06/install/include/Math/GenVector/PxPyPzM4D.h:108:14: note: in instantiation of member function 'ROOT::Math::PxPyPzE4D<double>::M' requested here
      fM = v.M();

```

In the header files all standard functions should be prefixed with `std::`. I guess this is a remnant from days when there were `using namespace std` statements in the headers...

Cheers,
Antoni

### PR DESCRIPTION
**Pull request by AdvaitDhingra**
This PR should solve [this](https://github.com/root-project/root/issues/7905) issue. (Closes #7905)

I looked for all of the standars library functions in the file, but std::sqrt() seemed to be the only one (not even std::cout was present).

I accidentally added another commit to this branch, but I reverted that, so this pr only fixed the named issue.

This *should* solve the issue, but let me know if it doesn't. 

Kind Regards,

Advait

### CODE REVIEW

**Comment by eguiraud** [comment]
@phsft-bot build

**Comment by lmoneta** [comment]
Thank you for this PR.
Since we include <cmath> I agree we should use `std::sqrt` instead simple `sqrt` that is defined in <math>

We should probably change other similar files in the genvector package. 

For this PR, it would be nice if you cleanup the commit history and squash the commits in a single one to have a cleaner history
Thank you

Lorenzo

**Comment by AdvaitDhingra** [comment]
> Thank you for this PR.
> Since we include I agree we should use `std::sqrt` instead simple `sqrt` that is defined in
> 
> We should probably change other similar files in the genvector package.
> 
> For this PR, it would be nice if you cleanup the commit history and squash the commits in a single one to have a cleaner history
> Thank you
> 
> Lorenzo

Sure, will do.

**Comment by AdvaitDhingra** [comment]
Oh god i did it again

**Comment by AdvaitDhingra** [comment]
That should be all of them

**Comment by lmoneta** [comment]
@phsft-bot build !

**Review by lmoneta — CHANGES_REQUESTED** [review]
Looks good to me. 
Thank you for adding also the other files in Genvector and the other functions. 

Can you just remove the two Cling commits ?

**Comment by AdvaitDhingra** [comment]
> Looks good to me.
> Thank you for adding also the other files in Genvector and the other functions.
> 
> Can you just remove the two Cling commits ?

Will do

**Comment by AdvaitDhingra** [comment]
There we go. The TCling commits are gone

**Comment by AdvaitDhingra** [comment]
> Build failed on ROOT-fedora31/noimt.
> Running on root-fedora-31-2.cern.ch:/home/sftnight/build/workspace/root-pullrequests-build
> [See console output](https://lcgapp-services.cern.ch/root-jenkins/job/root-pullrequests-build/114910/console).
> 
> ### Errors:
> * [2021-04-19T14:30:32.185Z] tor/Plane3D.h:239:27: error: no matching function for call to ‘sqrt(Vc_1::enable_if<true, Vc_1::Vector<double, Vc_1::VectorAbi::Sse> >)’
> * [2021-04-19T14:30:32.186Z] /usr/include/c++/9/cmath:475:5: error: no type named ‘__type’ in ‘struct __gnu_cxx::__enable_if<false, double>’
> * [2021-04-19T14:30:32.186Z] ests-build/build/include/Math/GenVector/Cartesian3D.h:113:39: error: no matching function for call to ‘sqrt(ROOT::Math::Cartesian3D<Vc_1::Vector<double, Vc_1::VectorAbi::Sse> >::Scalar)’
> * [2021-04-19T14:30:32.186Z] /usr/include/c++/9/cmath:475:5: error: no type named ‘__type’ in ‘struct __gnu_cxx::__enable_if<false, double>’

looks like I may have missed a few

**Comment by lmoneta** [comment]
Thank you !

The problem seems coming from using Vc with std::sort

**Comment by Axel-Naumann** [comment]
So that's because `std::sqrt` doesn't have an overload for Vc types. The solution is the following:
```
T SomeFunc() {
   using std::sqrt;
   return sqrt(fX);
}
```
i.e. keep `sqrt` unqualified such that ADL (argument-dependent lookup, fixing Vc) can happen, but make `std::sqrt` available through the local (!) `using namespace std::` or better yet `using std::sqrt` if that builds everywhere.

**Comment by AdvaitDhingra** [comment]
> So that's because `std::sqrt` doesn't have an overload for Vc types. The solution is the following:
> 
> ```c++
> T SomeFunc() {
>    using std::sqrt;
>    return sqrt(fX);
> }
> ```
> 
> i.e. keep `sqrt` unqualified such that ADL (argument-dependent lookup, fixing Vc) can happen, but make `std::sqrt` available through the local (!) `using namespace std::` or better yet `using std::sqrt` if that builds everywhere.

Ah ok. And just in the function(s) where the error is occurníng?

**Comment by AdvaitDhingra** [comment]
I misspellt the commit message. That was supposed to say "Vc sqrt error fix"

**Comment by AdvaitDhingra** [comment]
@phsft-bot build !

**Comment by Axel-Naumann** [comment]
Sorry - the same holds (of course...) for all the other `std::` functions: `sin`, `cos`,...: please use a `using` declaration.

**Comment by AdvaitDhingra** [comment]
> Sorry - the same holds (of course...) for all the other `std::` functions: `sin`, `cos`,...: please use a `using` declaration.

Ahh okay.

**Comment by AdvaitDhingra** [comment]
This should fix it now

**Comment by AdvaitDhingra** [comment]
Can you run @phsft-bot build ! again?

**Comment by lmoneta** [comment]
Thanks for the fixes. I'll run it now

**Comment by lmoneta** [comment]
@phsft-bot build !

**Comment by AdvaitDhingra** [comment]
Seems to have passed the tests :)

**Review by lmoneta — APPROVED** [review]
Looks good to me ! 

If it is fine for @Axel-Naumann , I will then merge it in master

Thank you for your contribution

**Comment by AdvaitDhingra** [comment]
> Looks good to me !
> 
> If it is fine for @Axel-Naumann , I will then merge it in master
> 
> Thank you for your contribution

Thank you for your help :)

File: math/genvector/inc/Math/GenVector/Cartesian2D.h [hunk 1]
```diff
@@ -89,7 +89,7 @@ public :
    Scalar X()     const { return fX;}
    Scalar Y()     const { return fY;}
    Scalar Mag2()  const { return fX*fX + fY*fY; }
-   Scalar R() const { return sqrt(Mag2()); }
+   Scalar R() const { using std::sqrt; return sqrt(Mag2()); }
    Scalar Phi() const { return (fX == Scalar(0) && fY == Scalar(0)) ? Scalar(0) : atan2(fY, fX); }
```
Comment (Axel-Naumann) [inline math/genvector/inc/Math/GenVector/Cartesian2D.h:93]: For a subsequent PR, `atan2` should get a similar treatment.

Comment (AdvaitDhingra) [reply math/genvector/inc/Math/GenVector/Cartesian2D.h:93]: Ah, ok

**Comment by AdvaitDhingra** [comment]
Okay, I've done the same for std::atan2 like so: 

´´´
void example() {
   using std::atan2;
   atan2(...)
} 
´´´

File: math/genvector/inc/Math/GenVector/eta.h [hunk 2]
```diff
@@ -52,8 +52,10 @@ namespace ROOT {
               static const Scalar big_z_scaled = pow(std::numeric_limits<Scalar>::epsilon(), static_cast<Scalar>(-.25));
 
               Scalar z_scaled = z/rho;
+              using std::fabs;
               if (std::fabs(z_scaled) < big_z_scaled) {
-                 return log(z_scaled + sqrt(z_scaled * z_scaled + 1.0));
+                 using std::sqrt;
+                 return log(z_scaled + std::sqrt(z_scaled * z_scaled + 1.0));
```
Comment (Axel-Naumann) [inline math/genvector/inc/Math/GenVector/eta.h:58]: ...and same for all occurrences of `log`. Sorry - I seem to find them only one by one :-/

**Review by Axel-Naumann — APPROVED** [review]
But even as is it's a nice improvement, LGTM!

File: math/genvector/inc/Math/GenVector/eta.h (hunk 2 above)
Comment (AdvaitDhingra) [reply math/genvector/inc/Math/GenVector/eta.h:58]: No problem. I'll change it :)

**Comment by AdvaitDhingra** [comment]
That should now be taken care of

**Comment by lmoneta** [comment]
@phsft-bot build

**Comment by AdvaitDhingra** [comment]
Oh no

**Comment by AdvaitDhingra** [comment]
Should I revert the std::log changes so the compiler doesn't run out of heap space?

**Comment by Axel-Naumann** [comment]
> Should I revert the std::log changes so the compiler doesn't run out of heap space?

No, please just ignore this. Let's wait for the other platforms to arrive and then if the other builds are green, merge.

**Comment by lmoneta** [comment]
Great ! We can then merge this. 
I will squash all commits in one to have a cleaner history
Thanks a lot for the contribution!

**Comment by AdvaitDhingra** [comment]
> Great ! We can then merge this.
> I will squash all commits in one to have a cleaner history
> Thanks a lot for the contribution!

Thank you for your help!

**Comment by AdvaitDhingra** [comment]
@lmoneta should I squash the commits, or do you want to do it?

**Comment by lmoneta** [comment]
I will do it. I will merge it no, thank you !

**Pull request merged**